src-plugins/mscAlgorithmPaint/mscAlgorithmPaintToolBox.cpp

```diff
@@ -1445,48 +1456,29 @@ void AlgorithmPaintToolBox::addSliceToStack(medAbstractView * view,const unsigne
             m_redoStacks->value(view)->clear();
 }
 
-void AlgorithmPaintToolBox::closeView()
+void AlgorithmPaintToolBox::clearToolBox()
 {
-    // close every data in the view
-    medAbstractView* view = qobject_cast<medAbstractView*>(QObject::sender());
-    medAbstractLayeredView* layeredView = dynamic_cast<medAbstractLayeredView*>(view);
-
-    if (layeredView->layersCount() > 0)
+    if (currentView && (currentView->layersCount()>0))
     {
-        // remove a data, which will recursively call this closeView() function
-        // until there is no more data in the view
-        layeredView->removeLayer(0);
-    }
-    else
-    {
-        // clean the toolbox when there is no data left in the view
-        onViewClosed(view);
-    }
-}
-
-void AlgorithmPaintToolBox::onViewClosed(medAbstractView* viewClosed)
-{
-    clearMask();
-
-    if (m_undoStacks->value(viewClosed))
-    {
-        m_undoStacks->value(viewClosed)->clear();
-        m_redoStacks->value(viewClosed)->clear();
-        m_undoStacks->remove(viewClosed);
-        m_redoStacks->remove(viewClosed);
+        // remove each "no name" mask data in the view
+        for(int unsigned i=0; i<currentView->layersCount(); i++)
+        {
+            if (currentView->layerData(i)->identifier() == "medImageMaskAnnotationData")
+            {
+                currentView->removeLayer(i); // call clearToolBox() again
+            }
+        }
+        clearMask(currentView);
+        updateView();
     }
-
-    showButtons(false);
-    resetToolbox();
-
-    if (viewClosed==currentView)
+    else // if no more data set in list, this is called a last time
     {
-        currentView = NULL;
-        m_itkMask = NULL;
+        clearMask(currentView);
+        updateView();
     }
 }
 
-void AlgorithmPaintToolBox::clearMask()
+void AlgorithmPaintToolBox::clearMask(medAbstractView* view)
```

Comment:
It's not clear why this function takes a medAbstractView now, since it is called with the current view. Did I miss something ?